esx_service/vmdk_ops.py

```diff
@@ -342,51 +347,109 @@ def detachVMDK(vmdk_path, vm_uuid):
     return disk_detach(vmdk_path, vm)
 
 
-# Check existence (and creates if needed) the path
-def getVolPath(vm_config_path):
-    # The folder for Docker volumes is created on <vm_datastore>/DOCK_VOLS_DIR
-    # On ESX the path is always /vmfs/volume/datastore/... , so we can do this:
-    vm_datastore = "/".join(vm_config_path.split("/")[0:4])
-    path = os.path.join(vm_datastore, DOCK_VOLS_DIR)
-
+# Check existence (and creates if needed) the path for docker volume VMDKs
+def get_vol_path(datastore):
+    # The folder for Docker volumes is created on <datastore>/DOCK_VOLS_DIR
+    path = os.path.join("/vmfs/volumes", datastore, DOCK_VOLS_DIR)
+    
     if os.path.isdir(path):
         # If the path exists then return it as is.
         logging.debug("Found %s, returning", path)
         return path
-
+    
     # The osfs tools are usable for all datastores
     cmd = "{0} {1}".format(OSFS_MKDIR_CMD, path)
     rc, out = RunCommand(cmd)
-    if rc != 0:
-        logging.warning("Failed to create %s", path)
-    else:
+    if rc == 0:
         logging.info("Created %s", path)
         return path
-
+    
+    logging.warning("Failed to create %s", path)
     return None
 
 
-def getVmdkName(path, vol_name):
-    # form full name as <path-to-volumes>/<volname>.vmdk
-    return os.path.join(path, "%s.vmdk" % vol_name)
+def known_datastores():
+    """returns names of know datastores"""
+    return [i[0] for i in vmdk_utils.get_datastores()]
 
+def parse_vol_name(full_vol_name):
+    """
+    Parses volume[@datastore] and returns (volume, datastore)
+    On parse errors raises ValidationError with syntax explanation
+    """
+    groups = re.match("\A([\w\-.]+)@?(\w+)?$", full_vol_name)
```

Comment:
So, with this change "@" has special meaning and can't be used in volume name. Update FAQ on wiki?

On a side note, docker volume create -d vmdk --name a/b` fails with
```
Error response from daemon: create a/b: VolumeDriver.Create: Failed to create /vmfs/volumes/vsan:521db3284facc204-03f387b708983c04/dockvols/a/b.vmdk. Failed to create virtual disk: 
```